test/test_utils.py

```diff
@@ -2392,6 +2392,33 @@ def test_traverse_obj(self):
         self.assertEqual(traverse_obj(etree, ('country', 0, ..., 'text()', {int_or_none})), [1, 2008, 141100],
                          msg='special transformations should act on current element')
 
+        self.assertEqual(traverse_obj(_TEST_DATA, [(100, 1.2), all]), [100, 1.2],
+                         msg='`all` should give all results as list')
+        self.assertEqual(traverse_obj(_TEST_DATA, [(100, 1.2), any]), 100,
+                         msg='`any` should give the first result')
+        self.assertEqual(traverse_obj(_TEST_DATA, [('dict', 'None', 100), all]), [100],
+                         msg='`all` should filter `None` and empty dict')
+        self.assertEqual(traverse_obj(_TEST_DATA, [('dict', 'None', 100), any]), 100,
+                         msg='`any` should filter `None` and empty dict')
+        self.assertEqual(traverse_obj(_TEST_DATA, [{
+                         'all': [('dict', 'None', 100, 1.2), all],
+                         'any': [('dict', 'None', 100, 1.2), any],
+                         }]), {'all': [100, 1.2], 'any': 100},
+                         msg='`all`/`any` should apply to each dict path separately')
+        self.assertEqual(traverse_obj(_TEST_DATA, [{
+                         'all': [('dict', 'None', 100, 1.2), all],
+                         'any': [('dict', 'None', 100, 1.2), any],
+                         }], get_all=False), {'all': [100, 1.2], 'any': 100},
+                         msg='`all`/`any` should apply to dict regardless of `get_all`')
+        self.assertEqual(traverse_obj(_TEST_DATA, [('dict', 'None', 100, 1.2), all, {float}]), None,
+                         msg='`all` should reset branching status')
+        self.assertEqual(traverse_obj(_TEST_DATA, [('dict', 'None', 100, 1.2), any, {float}]), None,
+                         msg='`any` should reset branching status')
```

Comment:
I don't understand this. Could you explain?

Shouldn't it be:

```
('dict', 'None', 100, 1.2) => [100, 1.2]
any => 100
{float} => 100
```

Reply:
nvm, `isinstance(100, float)` is not True. But this is yucky behavior imo, though off topic.

https://github.com/yt-dlp/yt-dlp/blob/14d0e23daf9a69e86fe5de4278184d7f13980741/yt_dlp/extractor/pr0gramm.py#L155 Luckily, this is the only code that uses `{float}`. But someone might easily mistake `{float}` to be sufficient here. Should/Can we do something about it?

Reply:
Actually, isn't that `pr0gramm` code buggy if site returns a mix of int and float? We'll yield all the `int`s first and then the `float`s. But order of `confidences` is important

Reply:
Bash proposed a `number` helper or `number_or_none`. Alternatively, we can support `{(int, float)}` as filter

Reply:
> Bash proposed a number helper or number_or_none

On discord? I can't find the convo.

Reply:
No just privately sometime back bc the int/float pattern came up before

Reply:
I like `{(int, float)}`, but doesn't really address the issue of accidentally using `{float}`.

We could make it a special case that accepts `int`? I don't see where we'd want to accept `float` but not `int`

Or, maybe it's not a big deal. I know I'd have made this mistake easily. But, we'd probably catch it in review?

Reply:
> I like `{(int, float)}`, but doesn't really address the issue of accidentally using `{float}`

People should think about type compatibility. I remember using `({int_or_none}, {float_or_none})` in crunchyroll so that it would try getting a whole number but show ints for specials

> We could make it a special case that accepts `int`? I don't see where we'd want to accept `float` but not `int`
That is true, since int is compatible with float (but not other way round). You could argue the same thing for using `int` when the type could be `float`, mainly related to javascript having only a single number type

> Or, maybe it's not a big deal. I know I'd have made this mistake easily. But, we'd probably catch it in review?

Id rather have people use an explicit `number` filter if they want to filter any number. Thinking about type compatibility should be encouraged.

Best to support both `{(type, type)}` as well as sth like `{number}` imo. I dont like having special cases anyways

Reply:
I'll note it for implementation in another PR